### ISSUE TITLE
NSInvocation not available in Swift – Needs replacement

### ISSUE THREAD

**Issue by fjaeger** [sparkle-project/Sparkle#1104]
I migrated an application from Objective-C to Swift and there is one major issue with Sparkle and using it from Swift: NSInvocation is not available in Swift and therefore, the following delegate methods cannot be implemented in my Swift-based AppDelegate:

```
- (BOOL)updater:(SUUpdater *)updater shouldPostponeRelaunchForUpdate:(SUAppcastItem *)item untilInvoking:(NSInvocation *)invocation;
- (void)updater:(SUUpdater *)updater willInstallUpdateOnQuit:(SUAppcastItem *)item immediateInstallationInvocation:(NSInvocation *)invocation;
```

Do you have any plans for a solution to this issue? Is there a workaround for now?

**Comment by pornel** [comment]
We could add new methods that take a block.

Alternatively, for Swift, change the header to use `SUInvocation`, and declare:

```objc
@interface SUInvocation 
- (void)invoke;
@end
```

Since it's dynamically dispatched the real class doesn't matter. It only needs to know that the `invoke` method exists.

**Comment by mminer** [comment]
Another workaround is to change the type of `invocation` to `UnsafeRawPointer`.

```swift
func updater(_ updater: SUUpdater, willInstallUpdateOnQuit item: SUAppcastItem, immediateInstallationInvocation invocation: UnsafeRawPointer) {
    // Terminate app
}
```

The compiler warns that the function nearly matches the real one, but Sparkle happily calls it.

**Comment by glouel** [comment]
Sorry to bounce an old issue but I'm trying to implement this in my Swift preference panel and I can't seem to get it to work. I think this may be because I'm using CocoaPods to include Sparkle. 

I did try both workarounds but none seem to work. Here's what I did :

```
final class PreferencesWindowController: NSWindowController, NSOutlineViewDataSource, NSOutlineViewDelegate, SUUpdaterDelegate {
```
Declare that we conform as SUUpdaterDelegate

```
    override func awakeFromNib() {
        super.awakeFromNib()
        sparkleUpdater = SUUpdater.init(for: Bundle(for: PreferencesWindowController.self))
        sparkleUpdater?.delegate = self
```
Add ourself as delegate


```
    func updater(_ updater: SUUpdater, willInstallUpdateOnQuit item: SUAppcastItem, immediateInstallationInvocation invocation: UnsafeRawPointer) {
        debugLog("this never gets called")
        self.close() // this is our own code to properly close our preference panel
    }
```

So with UnsafeRawPointer (second workaround), I do get the compiler warning about the near match and it doesn't get called. If I change it to SUInvocation as first suggested, and add the SUInvocation declaration in `SUUpdaterDelegate.h` and change the declaration to:
```
- (void)updater:(SUUpdater *)updater willInstallUpdateOnQuit:(SUAppcastItem *)item immediateInstallationInvocation:(SUInvocation *)invocation;
```

I don't get the near match compiler error (good) but it still doesn't get called. 

Am I right to assume that willInstallUpdateOnQuit should be triggered when I press the "Install and relaunch" button from Sparkle ? Maybe I'm missing something obvious and I apologize as I'm not really familiar with all of this !

**Comment by glouel** [comment]
Thanks for your very quick answer @kornelski, I'll make another issue then with what lead me to try that workaround.

**Comment by fjaeger** [comment]
Should be resolved by #1367
